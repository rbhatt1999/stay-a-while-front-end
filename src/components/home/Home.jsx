/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { BsFillStarFill } from 'react-icons/bs';
import ReactPaginate from 'react-paginate';
import { fetchHomeStays } from './HomeSlice';
import style from './Home.module.css';

const Home = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const token = localStorage.getItem('token');
  const [currentPage, setCurrentPage] = useState(0);
  const [totalPages, setTotalPages] = useState(0);
  const itemsPerPage = 6;
  const startIndex = currentPage * itemsPerPage;
  const endIndex = startIndex + itemsPerPage;
  const homeStayList = useSelector((state) => state.home.listings);

  useEffect(() => {
    dispatch(fetchHomeStays(token));
  }, [dispatch, token]);
  const handlePageChange = (selectedPage) => {
    setCurrentPage(selectedPage.selected);
  };

  useEffect(() => {
    if (homeStayList) {
      setTotalPages(Math.ceil(homeStayList.length / itemsPerPage));
    }
  }, [homeStayList]);

  const errMsg = useSelector((state) => state.home.error);
  const subset = homeStayList.slice(startIndex, endIndex);

  const handleClick = (id) => {
    navigate(`/home/${id}`);
  };

  if (homeStayList) {
    return (
      <div className={style.container}>
        <div className={style.cardContainer}>
          {subset.map((homeStay) => (
            <div
              key={homeStay.id}
              onClick={() => handleClick(homeStay.id)}
              onKeyDown={() => {}}
              role="button"
              tabIndex={homeStay.id}
              className={style.card}
            >
              <div className="card h-100">
                <div className="card-body">
                  {homeStay.images.length > 0
                  && (
                  <img
                    className={style.image}
                    src={homeStay.images[0].url}
                    alt={homeStay.name}
                  />
                  )}
                  <div className="d-flex justify-content-between mt-3">
                    <h4 className="f_size">{homeStay.location}</h4>
                    <h6 className="">
                      <BsFillStarFill />
                      {homeStay.rating}
                    </h6>
                  </div>
                  <h5 className="hse_name">{homeStay.name}</h5>
                </div>
              </div>
            </div>
          ))}
        </div>
        <ReactPaginate
          itemClass="pageitem"
          linkClass="pagelink"
          previousLabel="Prev"
          nextLabel="Next"
          pageCount={totalPages}
          onPageChange={handlePageChange}
          containerClassName="pagination justify-content-center"
          pageClassName="pageitem"
          pageLinkClassName="pagelink"
          activeClassName="active"
          previousClassName="pageitem"
          previousLinkClassName="pagelink"
          nextClassName="pageitem"
          nextLinkClassName="pagelink"
        />
      </div>
    );
  }

  return errMsg;
};
export default Home;
