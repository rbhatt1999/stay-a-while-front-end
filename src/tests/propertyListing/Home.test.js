import React from 'react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { Provider, useSelector } from 'react-redux';
import { render, act, screen } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import axios from 'axios';
import { BrowserRouter } from 'react-router-dom';
import homeSlice, { fetchHomeStays } from '../../components/home/HomeSlice';
import homestays from '../delete_homeStay/home_stays.json';
import Home from '../../components/home/Home';

jest.mock('axios');
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
}));
const store = configureStore({
  reducer: {
    home: homeSlice.reducer,
  },
});
describe('Home', () => {
  beforeEach(async () => {
    axios.get.mockResolvedValue({
      data: {
        data: {
          home_stays: homestays.data.home_stays,
        },
      },
    });
    useSelector.mockImplementation((selectorFn) => selectorFn({
      home: { listings: homestays.data.home_stays },
    }));
    await act(() => {
      store.dispatch(fetchHomeStays());
    });
    await act(async () => {
      render(
        <Provider store={store}>
          <BrowserRouter>
            <Home />
          </BrowserRouter>
        </Provider>,
      );
    });
  });
  test('should match screenshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <Home />
          </BrowserRouter>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('should have a image', () => {
    expect(screen.getAllByAltText('Home by Sam')).toHaveLength(1);
  });
});
