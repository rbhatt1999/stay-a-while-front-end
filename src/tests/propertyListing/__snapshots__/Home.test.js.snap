// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Show match snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="cardContainer"
    >
      <div
        class="card"
        role="button"
        tabindex="1"
      >
        <div
          class="card h-100"
        >
          <div
            class="card-body"
          >
            <img
              alt="Cottage by Maxwell"
              class="image"
              src="https://images.pexels.com/photos/2102587/pexels-photo-2102587.jpeg?auto=compress&cs=tinysrgb&w=600"
            />
            <div
              class="d-flex justify-content-between mt-3"
            >
              <h4
                class="f_size"
              >
                Sao Paulo, Brazil
              </h4>
              <h6
                class=""
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M3.612 15.443c-.386.198-.824-.149-.746-.592l.83-4.73L.173 6.765c-.329-.314-.158-.888.283-.95l4.898-.696L7.538.792c.197-.39.73-.39.927 0l2.184 4.327 4.898.696c.441.062.612.636.282.95l-3.522 3.356.83 4.73c.078.443-.36.79-.746.592L8 13.187l-4.389 2.256z"
                  />
                </svg>
                3
              </h6>
            </div>
            <h5
              class="hse_name"
            >
              Cottage by Maxwell
            </h5>
          </div>
        </div>
      </div>
    </div>
    <ul
      aria-label="Pagination"
      class="pagination justify-content-center"
      role="navigation"
    >
      <li
        class="pageitem disabled"
      >
        <a
          aria-disabled="true"
          aria-label="Previous page"
          class="pagelink "
          rel="prev"
          role="button"
          tabindex="-1"
        >
          Prev
        </a>
      </li>
      <li
        class="pageitem active"
      >
        <a
          aria-current="page"
          aria-label="Page 1 is your current page"
          class="pagelink"
          rel="canonical"
          role="button"
          tabindex="-1"
        >
          1
        </a>
      </li>
      <li
        class="pageitem disabled"
      >
        <a
          aria-disabled="true"
          aria-label="Next page"
          class="pagelink "
          rel="next"
          role="button"
          tabindex="-1"
        >
          Next
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Home should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="cardContainer"
    >
      <div
        class="card"
        role="button"
        tabindex="1"
      >
        <div
          class="card h-100"
        >
          <div
            class="card-body"
          >
            <img
              alt="Cottage by Maxwell"
              class="image"
              src="https://images.pexels.com/photos/2102587/pexels-photo-2102587.jpeg?auto=compress&cs=tinysrgb&w=600"
            />
            <div
              class="d-flex justify-content-between mt-3"
            >
              <h4
                class="f_size"
              >
                Sao Paulo, Brazil
              </h4>
              <h6
                class=""
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M3.612 15.443c-.386.198-.824-.149-.746-.592l.83-4.73L.173 6.765c-.329-.314-.158-.888.283-.95l4.898-.696L7.538.792c.197-.39.73-.39.927 0l2.184 4.327 4.898.696c.441.062.612.636.282.95l-3.522 3.356.83 4.73c.078.443-.36.79-.746.592L8 13.187l-4.389 2.256z"
                  />
                </svg>
                3
              </h6>
            </div>
            <h5
              class="hse_name"
            >
              Cottage by Maxwell
            </h5>
          </div>
        </div>
      </div>
    </div>
    <ul
      aria-label="Pagination"
      class="pagination justify-content-center"
      role="navigation"
    >
      <li
        class="pageitem disabled"
      >
        <a
          aria-disabled="true"
          aria-label="Previous page"
          class="pagelink "
          rel="prev"
          role="button"
          tabindex="-1"
        >
          Prev
        </a>
      </li>
      <li
        class="pageitem active"
      >
        <a
          aria-current="page"
          aria-label="Page 1 is your current page"
          class="pagelink"
          rel="canonical"
          role="button"
          tabindex="-1"
        >
          1
        </a>
      </li>
      <li
        class="pageitem disabled"
      >
        <a
          aria-disabled="true"
          aria-label="Next page"
          class="pagelink "
          rel="next"
          role="button"
          tabindex="-1"
        >
          Next
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
